src/viam/sdk/components/board/board.hpp

```diff
@@ -0,0 +1,162 @@
+    /// `servo` interrupt is configured this tracks the pulse width value. Consult Viam's `Board`
+    /// documentation for more details.
+    typedef int64_t digital_value;
+
+    /// @struct board_reader_status
+    /// @brief This contains all of the values for all of the registered analog readers and digital
+    /// interrupts that have been registered on the board.
+    struct status {
+        std::map<std::string, analog_value> analog_reader_values;
+        std::map<std::string, digital_value> digital_interrupt_values;
+    };
+
+    /// @enum power_mode
+    /// @brief Power mode of the board
+    /// The effect of these power modes depends on your physical board
+    enum class power_mode : uint8_t { normal = 0, offline_deep = 1 };
+
+    // functions shared across all components
+    static std::shared_ptr<ResourceSubtype> resource_subtype();
+    static Subtype subtype();
+
+    /// @brief Creates a `status` struct from its proto representation.
+    static status from_proto(viam::common::v1::BoardStatus proto);
+
+    /// @brief Creates a `analog_value` struct from its proto representation.
+    static analog_value from_proto(viam::common::v1::AnalogStatus proto);
+
+    /// @brief Creates a `digital_value` struct from its proto representation.
+    static digital_value from_proto(viam::common::v1::DigitalInterruptStatus proto);
+
+    /// @brief Creates a `power_mode` enum from its proto representation.
+    static power_mode from_proto(viam::component::board::v1::PowerMode proto);
+
+    /// @brief Converts a `status` struct to its proto representation.
+    static viam::common::v1::BoardStatus to_proto(status status);
+
+    /// @brief Converts a `analog_value` struct to its proto representation.
+    static viam::common::v1::AnalogStatus to_proto(analog_value analog_value);
+
+    /// @brief Converts a `digital_value` struct to its proto representation.
+    static viam::common::v1::DigitalInterruptStatus to_proto(digital_value digital_value);
+
+    /// @brief Converts a `power_mode` enum to its proto representation.
+    static viam::component::board::v1::PowerMode to_proto(power_mode power_mode);
+
+    /// @brief Get the status of all of the registered analog readers and digital interrupt readers
+    virtual status get_status() = 0;
+
+    /// @brief Gets the high/low state of the given pin on a board.
+    /// @param pin board pin name
+    /// @return high/low state of the given pin. High = on, low = off
+    virtual bool get_gpio(const std::string& pin) = 0;
+
+    /// @brief Set the gpio high/low state of the given pin on a board
+    /// @param high true if the pin should be set to high (on) or false if it should be low (off)
+    virtual void set_gpio(const std::string& pin, bool high) = 0;
+
+    /// @brief Gets the duty cycle of the given pin on a board.
+    /// @param pin board pin name
+    /// @return duty cycle percentage (0 to 1)
+    virtual double get_pwm_duty_cycle(const std::string& pin) = 0;
```

Comment:
The other sdks call this function pair `get_pwm` and `set_pwm`. I found this confusing because we also have `get_pwm_frequency` and `set_pwm_frequency` and it is not obvious (to me) that `set_pwm` would refer to the duty cycle rather than to the frequency.